### ISSUE TITLE
[Bug] 

### ISSUE THREAD

**Issue by wutianxu** [open-mmlab/mmdetection#9484]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmdetection/issues) and [Discussions](https://github.com/open-mmlab/mmdetection/discussions) but cannot get the expected help.
- [X] I have read the [FAQ documentation](https://mmdetection.readthedocs.io/en/latest/faq.html) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version (master)](https://github.com/open-mmlab/mmdetection) or [latest version (3.x)](https://github.com/open-mmlab/mmdetection/tree/dev-3.x).

### Task

I have modified the scripts/configs, or I'm working on my own tasks/models/datasets.

### Branch

master branch https://github.com/open-mmlab/mmdetection

### Environment

sys.platform: linux
Python: 3.9.13 (main, Aug 25 2022, 23:26:10) [GCC 11.2.0]
CUDA available: True
GPU 0: NVIDIA GeForce RTX 3090 Ti
CUDA_HOME: /home/tianxu/anaconda3/envs/pytorch
NVCC: Cuda compilation tools, release 11.7, V11.7.99
GCC: gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
PyTorch: 1.13.0
PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.6.0 (Git Hash 52b5f107dd9cf10910aaa19cb47f3abf9b349815)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.7
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.5
  - Magma 2.6.1
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.7, CUDNN_VERSION=8.5.0, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -fabi-version=11 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wunused-local-typedefs -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.13.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

TorchVision: 0.14.0
OpenCV: 4.6.0
MMCV: 1.7.0
MMCV Compiler: GCC 9.3
MMCV CUDA Compiler: 11.7
MMDetection: 2.26.0+762a39d

### Reproduces the problem - code sample

mmdet/models/necks/fpn.py
```

def forward(self, inputs):
        """Forward function."""
        assert len(inputs) == len(self.in_channels)

        # build laterals
        laterals = [
            lateral_conv(inputs[i + self.start_level])
            for i, lateral_conv in enumerate(self.lateral_convs)
        ]

        # build top-down path
        used_backbone_levels = len(laterals)
        for i in range(used_backbone_levels - 1, 0, -1):
            # In some cases, fixing `scale factor` (e.g. 2) is preferred, but
            #  it cannot co-exist with `size` in `F.interpolate`.
            if 'scale_factor' in self.upsample_cfg:
                # fix runtime error of "+=" inplace operation in PyTorch 1.10
                laterals[i - 1] = laterals[i - 1] + F.interpolate(
                    laterals[i], **self.upsample_cfg)
            else:
                prev_shape = laterals[i - 1].shape[2:]
                laterals[i - 1] = laterals[i - 1] + F.interpolate(
                    laterals[i], size=prev_shape, **self.upsample_cfg)

        # build outputs
        # part 1: from original levels
        outs = [
            self.fpn_convs[i](laterals[i]) for i in range(used_backbone_levels)
        ]
        # part 2: add extra levels
        if self.num_outs > len(outs):
            # use max pool to get more levels on top of outputs
            # (e.g., Faster R-CNN, Mask R-CNN)
            if not self.add_extra_convs:
                for i in range(self.num_outs - used_backbone_levels):
                    outs.append(F.max_pool2d(outs[-1], 1, stride=2))
            # add conv layers on top of original feature maps (RetinaNet)
            else:
                if self.add_extra_convs == 'on_input':
                    extra_source = inputs[self.backbone_end_level - 1]
                elif self.add_extra_convs == 'on_lateral':
                    extra_source = laterals[-1]
                elif self.add_extra_convs == 'on_output':
                    extra_source = outs[-1]
                else:
                    raise NotImplementedError
                outs.append(self.fpn_convs[used_backbone_levels](extra_source))
                for i in range(used_backbone_levels + 1, self.num_outs):
                    if self.relu_before_extra_convs:
                        outs.append(self.fpn_convs[i](F.relu(outs[-1])))
                    else:
                        outs.append(self.fpn_convs[i](outs[-1]))
        # concat and split 
        feat_flatten = [out.flatten(2) for out in outs]     # outs: list( [BN, channel, width, height] )
                                                            # feat_flatten [BN, channel, sum(width*height)]
        spatial_shapes = [out.shape[2:4] for out in outs] 
        
        feat_flatten = torch.cat(feat_flatten, 2)
        bn,channel,_ = feat_flatten.shape
        value_list = feat_flatten.split([H_ * W_ for H_, W_ in spatial_shapes],
                            dim=2)
        if torch.isnan(feat_flatten).any():
            a = 1
        if torch.isinf(feat_flatten).any():
            a = 1
        out_results = [torch.reshape(value_list[i],[bn, channel, spatial_shapes[i][0], spatial_shapes[i][1]] ) for i in range(len(value_list))]
        return tuple(out_results )
```

### Reproduces the problem - command or script

python tools/train.py --config configs/faster_rcnn/faster_rcnn_r101_fpn_1x_voc.py

### Reproduces the problem - error message

Orignal training results:
2022-12-05 22:16:31,651 - mmdet - INFO - Saving checkpoint at 12 epochs
2022-12-05 22:18:55,627 - mmdet - INFO - 
+-------------+------+-------+--------+-------+
| class       | gts  | dets  | recall | ap    |
+-------------+------+-------+--------+-------+
| aeroplane   | 285  | 1140  | 0.547  | 0.334 |
| bicycle     | 337  | 1578  | 0.816  | 0.557 |
| bird        | 459  | 1225  | 0.516  | 0.323 |
| boat        | 263  | 1598  | 0.456  | 0.178 |
| bottle      | 469  | 1515  | 0.416  | 0.212 |
| bus         | 213  | 1305  | 0.765  | 0.539 |
| car         | 1201 | 3848  | 0.721  | 0.533 |
| cat         | 358  | 2484  | 0.846  | 0.525 |
| chair       | 756  | 3991  | 0.570  | 0.316 |
| cow         | 244  | 2103  | 0.836  | 0.414 |
| diningtable | 206  | 2860  | 0.811  | 0.379 |
| dog         | 489  | 3100  | 0.855  | 0.429 |
| horse       | 348  | 2011  | 0.853  | 0.465 |
| motorbike   | 325  | 2194  | 0.720  | 0.439 |
| person      | 4528 | 16298 | 0.811  | 0.593 |
| pottedplant | 480  | 1751  | 0.346  | 0.184 |
| sheep       | 242  | 2280  | 0.744  | 0.368 |
| sofa        | 239  | 3136  | 0.820  | 0.373 |
| train       | 282  | 2312  | 0.837  | 0.565 |
| tvmonitor   | 308  | 1217  | 0.721  | 0.540 |
+-------------+------+-------+--------+-------+
| mAP         |      |       |        | 0.413 |
+-------------+------+-------+--------+-------+
2022-12-05 22:18:55,672 - mmdet - INFO - Exp name: faster_rcnn_r101_fpn_1x_voc.py
2022-12-05 22:18:55,672 - mmdet - INFO - Epoch(val) [12][4952]	mAP: 0.4133, AP50: 0.4130

The modified trainning results:
---------------iou_thr: 0.5---------------
2022-12-13 14:19:50,028 - mmdet - INFO - 
+-------------+-----+------+--------+-------+
| class       | gts | dets | recall | ap    |
+-------------+-----+------+--------+-------+
| aeroplane   | 39  | 0    | 0.000  | 0.000 |
| bicycle     | 34  | 0    | 0.000  | 0.000 |
| bird        | 47  | 0    | 0.000  | 0.000 |
| boat        | 47  | 0    | 0.000  | 0.000 |
| bottle      | 38  | 0    | 0.000  | 0.000 |
| bus         | 21  | 0    | 0.000  | 0.000 |
| car         | 134 | 0    | 0.000  | 0.000 |
| cat         | 37  | 0    | 0.000  | 0.000 |
| chair       | 106 | 0    | 0.000  | 0.000 |
| cow         | 29  | 0    | 0.000  | 0.000 |
| diningtable | 29  | 0    | 0.000  | 0.000 |
| dog         | 59  | 0    | 0.000  | 0.000 |
| horse       | 41  | 0    | 0.000  | 0.000 |
| motorbike   | 31  | 0    | 0.000  | 0.000 |
| person      | 555 | 0    | 0.000  | 0.000 |
| pottedplant | 48  | 0    | 0.000  | 0.000 |
| sheep       | 26  | 0    | 0.000  | 0.000 |
| sofa        | 27  | 0    | 0.000  | 0.000 |
| train       | 38  | 0    | 0.000  | 0.000 |
| tvmonitor   | 36  | 0    | 0.000  | 0.000 |
+-------------+-----+------+--------+-------+
| mAP         |     |      |        | 0.000 |
+-------------+-----+------+--------+-------+
2022-12-13 14:19:50,070 - mmdet - INFO - Exp name: faster_rcnn_r101_fpn_1x_voc.py
2022-12-13 14:19:50,070 - mmdet - INFO - Epoch(val) [12][600]   mAP: 0.0000, AP50: 0.0000

### Additional information

I am trying to add self-attention to the Faster-RCNN network, so the output features of FPN were flattened in the dimension  of width and height, then concatenate the results in the flattened dimension. After that, I just split the concat feature and reshape these separated features to fit the shape of output features from FPN. I have checked that the new output features were equal to the  original output from FPN. The strange thing is that the model is not convergent, and the training accuracy is 0, while the original model works well.

The dataset was generated from VOC2007, which contains 10 images of each instance.

**Comment by ZwwWayne** [comment]
You should check whether the implementation of your module is wrong.

**Comment by wutianxu** [comment]
> You should check whether the implementation of your module is wrong.

As I mentioned, the output features of modified FPN were equal to the original FPN network, which has been checked by Torch.Tensor.equal:

```
for idx in range(len(outs)):
	print(outs[idx].euqal(out_results[idx]))
```

By the way, the code have been checked on detectron2, it works well.

```python
    feat_flatten = [out.flatten(2) for out in outs]     # outs: list( [BN, channel, width, height] )
                                                        # feat_flatten [BN, channel, sum(width*height)]
    spatial_shapes = [out.shape[2:4] for out in outs] 
    
    feat_flatten = torch.cat(feat_flatten, 2)
    bn,channel,_ = feat_flatten.shape
    value_list = feat_flatten.split([H_ * W_ for H_, W_ in spatial_shapes],
                        dim=2)
    if torch.isnan(feat_flatten).any():
        a = 1
    if torch.isinf(feat_flatten).any():
        a = 1
    out_results = [torch.reshape(value_list[i],[bn, channel, spatial_shapes[i][0], spatial_shapes[i][1]] ) for i in range(len(value_list))]
    return tuple(out_results )
```
